### PR TITLE
Reflection, Query Builders and Scope Creep

### LINKED ISSUE #19: Add more control for adding fields to QueryBuilder's queries
The use of List<Schema.SObjectField> & field sets is helpful, but with the current implementation, you can only choose 1 of 3 options (based on the values provided to QueryBuilder's constructor)
1. Only SObjectType provided - all fields are included in the query
2. Field Set is provided - only the fields in the field set are included in the query
3. List of SObjectField is provided - only the fields in the list set are included in the query

There is currently no way to leverage a mixture of these options. To provided more flexibility, I think there should be new methods added to QueryBuilder.cls
- public IQueryBuilder addFields(Schema.FieldSet fieldSet)
- public IQueryBuilder addFields(List<Schema.SObjectField> fields)
- public IQueryBuilder addAllStandardFields()
- public IQueryBuilder addAllCustomFields()

This would allow chaining these methods to give granular control over what fields are included in the resulting query, like...
```
return this.Query
    .addAllStandardFields() // Auto-add all standard fields provided by Salesforce
    .addFields(myFieldSet) // Add all fields in the field set
    .addFields(myListOfSObjectFields) //Add all fields in the list
    .filterBy(queryFilters)
    .getQueryResults();
```

In addition, the Repository Settings (API name NebulaRepositorySettings__c) could have new settings added
1. IncludeCommonFields__c - **existing**. Adds common fields like Id, OwnerId, Name, Subject, CreatedById, etc if they exist on the current SObject Type
2. IncludeStandardFields__c - **new**. Overlaps with IncludeCommonFields__c a bit. This would add all standard fields to the query
2. IncludeCustomFields__c - **new**. This would add all custom fields to the query

The main issue with adding the custom settings is that the custom settings are not specific to a particular SObject - if you were to use the proposed IncludeStandardFields__c setting, with the current architecture, all queries would get standard fields. This could be fine for some orgs/situations, but need to evaluate if adding these fields to the repo settings makes sense.

### LINKED ISSUE #36: Ability to query with OR statements

### LINKED ISSUE #37: Support for aggregate queries
I may work on this piece after I finish up work on #13

### LINKED ISSUE #50: Add more control for adding parent fields to a query
SObjectRepository & QueryBuilder now have a basic way to include all fields in a query, but when used, fields from parent objects cannot be included. There should be an additional method added that allows including all fields from the base object while also specifying a field set, a list of sobject fields, or both.

### LINKED ISSUE #52: Add the ability to filter children object queries

### LINKED ISSUE #68: Use reflection when possible throughout the framework

### LINKED ISSUE #69: Setup ApexDoc (JavaDoc for Apex)
There's a project called [ApexDoc](https://github.com/SalesforceFoundation/ApexDoc) that can be used to generate documentation in Apex classes. Since Nebula v2.0 is going to break a lot of things, I want to see if using ApexDoc would be useful.

### PR DESCRIPTION
**Pull request by jongpie**
What started out as a journey to implement reflection in Nebula... turned into a journey of reflecting inward... and also, I decided to change a lot of stuff.

Fixes #19, fixes #36, fixes #37, fixes #50, fixes #52, fixes #68, fixes #69

Draft of release notes are below - I'll update them here as I wrap up the pull request.
_________________________

## NebulaFactory.cls
The framework now supports reflection for some interfaces. To utilise your own interface implementation, open the custom metadata type 'Nebula Interface Implementations', find the record of the desired interface and enter the name of your Apex class. Nebula will then inject your code into the framework at runtime. 
        1. IAggregateResultQueryBuilder
        2. IQueryArgumentFormatter
        3. IQueryFilter
        4. ISearchQueryBuilder
        5. ISObjectQueryBuilder

To leverage Apex's implementation of reflection, all relevant classes now use a parameterless constructor - any setup work needed has been moved from constructors to methods.

# SObject Repository
## DML.cls
* All DML methods now return the corresponding list of database results - see IDML.cls. Previously, these were void methods. This change is inherited by SObjectRepository.cls as well (it extends DML.cls)
## SObjectRepository.cls
* The member variable sobjectRepository.Query is now an instance of SObjectQueryBuilder.cls - see '*Query Builders*' section for more info
- New member variable sobjectRepository.AggregateQuery, an instance of the new class AggregateResultQueryBuilder.cls - see '*Query Builders*' section for more info

# Query Builders
QueryBuilder.cls has been split into SObjectQueryBuilder.cls, AggregateQueryBuilder.cls & SearchQueryBuilder.cls

## QueryBuilder.cls
* QueryBuilder.cls is now an abstract class. It only contains protected methods for shared logic that is used by SObjectQueryBuilder.cls & AggregateQueryBuilder.cls. QueryBuilder.cls itself should no longer be used directly.

## SObjectQueryBuilder.cls
A new builder class that generates dynamic SOQL queries & returns a list of SObject
* 7 ways to add fields to your query: 
        1. addFields(Schema.FieldSet fieldSet)
        2. addFields(List<Schema.SObjectField> fields)
        3. addAllFields()
        4. addAllStandardFields()
        5. addAllCustomFields()
        6. addAllReadableFields()
        7. addAllEditableFields()
* Name fields for lookups are automatically added to your query. For example, adding the contact object's field AccountId (a lookup to account):
    ```sobjectQueryBuilder.addField(Schema.Contact.AccountId)```

     will result in your query having 2 fields added 
     ```SELECT AccountId, Account.Name FROM Contact```
* New method for including children records
* New method, getQueryLocator()

## AggregateQueryBuilder.cls
A new builder class that generates dynamic SOQL queries & returns a list of AggregateResult
* Four methods to add fields to group by 
        1. groupBy(QueryDate queryDate);
        2. groupBy(Schema.FieldSet fieldSet);
        3. groupBy(Schema.SObjectField groupByField);
        4. groupBy(List<Schema.SObjectField> groupByFields);

* Six aggregate methods
        1. avg(Schema.SObjectField sbojectField)
        2. count(Schema.SObjectField sbojectField)
        3. count_distinct(Schema.SObjectField sbojectField)
        4. max(Schema.SObjectField sbojectField)
        5. min(Schema.SObjectField sbojectField)
        6. sum(Schema.SObjectField sbojectField)

Since this is the first round of implementing the class, not all aggregate result features (like 'GROUP BY ROLLUP') have been implemented, but more will be added in the future.

## SearchQueryBuilder.cls
A new builder class that generates dynamic SOSL queries & returns a list of a list of SObject (yes, a list of list, that's not a typo)
* Add SObject Query Builders to include them in the returned search results
        1.  addSObjectQueryBuilders(ISObjectQueryBuilder sobjectQueryBuilder) - add a single SObject query builder
        2.  addSObjectQueryBuilders(List<ISObjectQueryBuilder> sobjectQueryBuilders) - add a list of SObject query builders
* The query search group defaults to 'ALL FIELDS', but can be changed to any of the QuerySearchGroup enums
        1.  setQuerySearchGroup(QuerySearchGroup searchGroup)

## QueryFilter.cls
* Completely changed the methods available to handle 4 filtering scenarios
        1. filterByField
        2. filterByParentField
        3. filterByQueryDate
        4. filterBySubquery
* Added support for specifying 'AND' and 'OR' statements in a query. To use, provide a list of query filters that you want to be wrapped as a set of 'AND' or 'OR' statements to the corresponding method
        1. andFilterBy
        2. orFilterBy

## QueryDate.cls
* A new class that can be used in two ways
        1. Filter on a date/datetime field, using date functions
```
QueryDate qd = QueryDate.CALENDAR_MONTH(Schema.Lead.CreatedDate);
QueryFilter filter = new QueryFilter().filterByQueryDate(qd, QueryOperator.EQUALS, 3);
sobjectQueryBuilder.filterBy(filter);
System.assertEquals(filter.getValue(), 'CALENDAR_MONTH(CreatedDate) = 3');

// In SOQL, it would be written as
List<Lead> leads = [SELECT Id
FROM Lead
WHERE CALENDAR_MONTH(CreatedDate) = 3')];
```

   2. Group by a date function in an aggregate result

```
QueryDate qd = QueryDate.CALENDAR_MONTH(Schema.Lead.CreatedDate);
aggregateResultQueryBuilder.groupBy(qd);

// In SOQL, it would be written as
List<AggregateResult> results = [SELECT CALENDAR_MONTH(CreatedDate)
FROM Lead
GROUP BY CALENDAR_MONTH(CreatedDate)];
```

## QueryFilterScope.cls
* Added 3 new filter scope enums
        1. DELEGATED
        2. MY_TERRITORY - only works in orgs with territory management enabled
        3. MY_TEAM_TERRITORY - only works in orgs with territory management enabled

# Record Types
## SObjectRecordTypes.cls
* Now requires the abstract method `Schema.SObjectType getSObjectType()` to be implemented
# Settings
* The custom setting NebulaRepositorySettings__c.object has been renamed to NebulaSObjectQueryBuilderSettings__c.object

# Utilities
## CollectionUtils.cls
* Added 3 new methods for working with collections
        1. getLastItem
        2. pop
        3. splice

## Environment.cls
* New property, NamespacePrefix

# Documentation
* Started commenting classes & methods with [ApexDoc](https://github.com/SalesforceFoundation/ApexDoc). This will be an ongoing improvement.

### CODE REVIEW

**Comment by jongpie** [comment]
@jamessimone I'm still working on wrapping up some changes for this pull request (and catching up on tests, release notes, etc), but if you're bored and want to take a look, this has most of the changes that I'm making for v2.0

**Comment by jongpie** [comment]
I've added `SearchQueryBuilder.cls`, giving a total of 3 types of query builders now
  1. SearchQueryBuilder - SOSL queries
  2. SObjectQueryBuilder - standard SOQL queries
  3. AggregateResultQueryBuilder - SOQL queries for aggregate results

**Comment by jamessimone** [comment]
@jongpie I began giving this an in-depth look yesterday, as I wanted to port the AggregateResultQueryBuilder over to our code.  I ended up taking a slightly different approach with a few of the design choices that you made - for example, I had the AggregateResultQueryBuilder extending QueryBuilder without a separate interface.  This made some of the method calls slightly clearer.  I also am not sure that I see the value in adding reflection for the QueryBuilder implementations - it's hard to imagine somebody wanting to override the default implementations (I would argue that the defaults are actually the reason somebody would want to use Nebula to begin with - nobody wants to have to re-write the query-level code!), and the work necessary to support reflection (specifically, moving the setting of the SObjectType for each QueryBuillder out of the constructors) ends up leading to additional hoops to jump through.  Does that make sense?

I also wrapped the implementation for grabbing the chained string for relationship names to another object outside of QueryFilter - the reason for this is that we have a concrete implementation right now that uses the orderBy call in QueryBuilder on a parent field (CustomObject__c.Lead__r.CreatedDate).  So I changed the implementation for IQueryBuilder's interface to accept this object, SObjectFieldWrapper, and that object handlers either a single SObjectField, or a chained list of related SObjectFields.  

This SObjectFieldWrapper I also used in the QueryFilter implementation so that there is only one "filterByField" method, so to speak; the first argument is replaced by the wrapper.  More than happy to drop that in a branch if you are interested in taking a look.

I need to look into the rest of the changes you made today, and I am sure you plan to make others as well.  Overall I have to say that the quality of the code is quite high, and aside from the stuff that I've mentioned (which is very specific in terms of feedback), it has been a pleasure to look through all of this!

**Comment by jongpie** [comment]
@jamessimone your comments about reflection for the query builders totally makes sense - I started working on just implementing reflection, but query builder (the one class at the time) was giving me some problems when trying to convert to work with reflection, plus I knew I wanted to eventually make some enhancements to QueryBuilder.cls anyway, and well... I went down a rabbit hole of refactoring and changing. In the end, I think you're right - I don't think there's a benefit to making the query builder classes use reflection. I'm going to undo some of the reflection-related changes that I made and move some of the code back into the constructors. 

Your SObjectFieldWrapper class sounds like a great idea, that might actually help with a few other things that I want to add to the query builder classes in the future - if you're cool with throwing it into a branch, I'd love to take a look.

Last thing to note - I added SearchQueryBuilder.cls for SOSL queries. I'm not sure if you saw that change already or not, but it involved a few more changes to SObjectQueryBuilder.cls

**Comment by jamessimone** [comment]
@jongpie I will throw that in a branch, then!  Also, two other notes: 
*  the getFirstQueryResult / getQueryResult calls in AggregateResultQueryBuilder can be eliminated entirely (!)
* I don't know if you will want to follow this approach, but I changed the aggregate methods in AggregateResultQueryBuilder to take in a string as well as the SObjectField, with the string being the alias.  I get what you were going for with the default names, and it would be nice to have that as a fallback (though with method overloading that's gonna grow the class by quite a bit) - we have production code already referencing distinct aliases in the aggregate results, and it was more convenient for me to just require an alias than it was to do any other thing.  Your thoughts?

**Comment by jongpie** [comment]
@jamessimone couple of notes for you!
* My brain is fried, so I'm sure I'm overlooking something - why can the getFirstQueryResult / getQueryResult calls in AggregateResultQueryBuilder be eliminated?
* I definitely plan to add the ability to alias the aggregate functions, but I personally will end up using the auto-alias names, so I want to support both options. I originally started implementing 2 versions for each aggregate function (see below), but then I wanted to see if there was a cleaner way. So far, overloading still seems like the best option for supporting both
    * Auto-aliased: `avg(Schema.SObjectField numericSObjectField)`
    * Manually aliased: `avg(Schema.SObjectField numericSObjectField, String fieldAlias)`
* I've renamed SObjectFieldWrapper to QueryField (and for now, I'm sticking with your implementation of passing SObject fields in the constructor - seems to be the shortest code when using QueryField). It's now being used in these public methods (and probably a few other classes that I'm forgetting + test classes)
  * AggregateResultQueryBuilder - groupBy() and orderBy() now use QueryField
  * SObjectQueryBuilder - orderBy() now uses QueryField
  * QueryFilter - you changed this already for SObjectFieldWrapper, I just changed it to the new class name, QueryField 

**So, now we can group by, filter on and order by parent fields, which is pretty freakin' sweet. Thanks again for the help!**

I'm still cleaning up the code, catching up on tests and stripping out some of the reflection changes that I implemented, so still a bit more work for me to do on this PR, but keep the feedback coming if you have more ideas.

**Comment by jongpie** [comment]
@jamessimone welp, I made 2 more changes...

1. I had previously forgotten to update the aggregate functions in AggregateQueryBuilder.cls to use QueryField instead of Schema.SObjectField. That's done, so now you could do something like this (my examples aren't fully working code, just snippets of the relevant bits, let me know if they're unclear though)
```
// Get the min created date of the parent object (account)
[SELECT MIN(Account.CreatedDate) FROM Contact];
```
by doing something like this
```
QueryField parentAcctCreatedDate = new QueryField(new List<SObjectField>(Contact.AccountId, Account.CreatedDate));
// Get the min created date of the parent object (account)
aggregateQueryBuilder.min(parentAcctCreatedDate);
```

2. All the aggregate functions for AggregateQueryBuilder have been overloaded so you can supply your own field alias - just for you, dawg. Example:
```
// Automatic field alias
aggregateResultQueryBuilder.avg(myQueryField);
// Bring Your Own Field Alias (BYOFA)
aggregateResultQueryBuilder.avg(myQueryField, 'wubbaLubbaDubDub');

**Comment by jamessimone** [comment]
@jongpie due to the magical mysteries of Apex, you don't even need to cast a `List<SObject>` to a `List<AggregateResult>` - so your base `getQueryResults()` and `getFirstQueryResult()` methods can be called without needing specific methods to get the aggregate results.

ALSO - I ended up making the 'getQuery()' method on QueryBuilder virtual and added the following to the Aggregate's over-ridden implementation:
`if(this.aggregateFunctionList.isEmpty()) {'
            return super.getQuery();'
        `}`  (sorry there's some weird formatting there with the code snippet, but I'm sure you get the point)

That way if you have repositories overriding the default query builder, you can still do regular queries so long as you aren't doing anything with aggregates.

Grouping by parent fields is pretty awesome (as are those BYOA's) - thank YOU for adding those in!

**Comment by jongpie** [comment]
@jamessimone I think I'm at a stopping point now. Do you mind taking another glance at the changes? Test classes are still lacking, but the main classes should be good now. I've essentially removed all of the reflection bits, even though that was the original aim. Since I ended up refactoring so much other stuff and new adding query builders, etc, I don't think there's now much value in supporting reflection right now. It may get added back in the future, but I prefer using constructors for some parameters. 

I also want your opinion on QueryFilter in particular - it's still using a parameterless constructor, with separate methods for different filtering scenarios. I want to try to move back to passing parameters in the constructor - essentially, each 'filterby<something>' method would get converted to a constructor. Thoughts? Should I move back to parameters in the constructors? Or keep the methods as-is?

**Comment by jamessimone** [comment]
@jongpie I will take a look!

**Comment by jongpie** [comment]
@jamessimone please don't hate me, but I just pushed one last set of changes for SearchQueryBuilder.cls. I realized this weekend that I want all 3 query builders to be usable as soon as they are constructed - meaning, as soon as you construct one, you should be able to call getQueryResults() and not receive a runtime error. In the real world, the query builders would also need filters added, query fields to be selected, etc but I think it's nice to have the base object be usable.

SearchQueryBuilder.cls didn't work this way - you had to call a few of the builder methods before the search query would be usable. I've changed the constructors in SearchQueryBuilder. As a result, SObjectRepository.cls no longer has the query builder variables (this.Query, this.AggregateResult and this.SearchResult) - instead, each method in SObjectRepo now just instantiates it's own query builder as needed.

AggregateResultQueryBuilder.cls had a small change added - if you run something like this (no 'group by' or aggregate functions specified), then a default 'COUNT(Id)' is added to the aggregate query.
`new AggregateResultQueryBuilder(Schema.Account.SObjectType).getQueryResults();`

**Comment by jamessimone** [comment]
@jongpie no hate, and no worries!  Reviewing the rest of your changes (I am mostly caught up) was one of my plans this weekend - sorry I spent it going over Lightning stuff instead.  I plan to take a look and get this fully reviewed by the end of the week.

File: src/classes/SObjectRecordTypes.cls [hunk 1]
```diff
@@ -48,19 +48,19 @@ public abstract class SObjectRecordTypes extends NebulaCore implements ISObjectR
     private void populateCache() {
         if(cachedRecordTypesBySObjectMap.containsKey(this.sobjectName)) return;
 
-        // Trying to call Schema.RecordType.SObjectType confuses Apex, so we have to get it do an extra describe call
+        // Trying to call Schema.RecordType.SObjectType confuses Apex, so we have to get it through an extra describe call
```
Comment (jamessimone) [inline src/classes/SObjectRecordTypes.cls]: that is kind of hilarious

File: src/classes/SObjectRepository.cls [hunk 2]
```diff
@@ -50,14 +27,14 @@ public abstract class SObjectRepository extends DML implements ISObjectRepositor
 
     // SOQL
     public virtual SObject getById(Id recordId) {
-        return this.Query
+        return new SObjectQueryBuilder(this.getSObjectType())
```
Comment (jamessimone) [inline src/classes/SObjectRepository.cls]: i really like this change!

File: src/classes/AggregateResultQueryBuilder.cls [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+            + '\nFROM ' + this.sobjectType.getDescribe().getName()
+            + super.doGetWhereClauseString()
+            + this.getGroupByFieldString(true)
+            + super.doGetOrderByString()
+            + super.doGetLimitCountString();
+
+        return queryString;
+    }
+
+    public AggregateResult getFirstQueryResult() {
```
Comment (jamessimone) [inline src/classes/AggregateResultQueryBuilder.cls:154]: did you try taking these out?  https://developer.salesforce.com/docs/atlas.en-us.api.meta/api/sforce_api_calls_query_aggregateresult.htm - aggregate result is just a read-only SObject

**Review by jamessimone — APPROVED** [review]
Pretty much looks good man.  I think QueryDate, QueryField, and the new QueryBuilders are great additions.  I left the one comment about the aggregate results, but you may feel that having those methods there increases clarity.  Either way, thumb's up!

File: src/classes/AggregateResultQueryBuilder.cls (hunk 3 above)
Comment (jongpie) [reply src/classes/AggregateResultQueryBuilder.cls:154]: Yo! Thanks for reviewing! I think I'm going to keep these methoda for now, just for the sake of clarity and being explicit when possible

**Comment by jongpie** [comment]
@jamessimone thanks for the review! I'm going to keep those methods for aggregate results for now. I plan to merge this to dev and master later today.

**Pull request merged**